Data obtained from `https://data.boston.gov/dataset/buildbps-facilities-and-educational-data-for-boston-public-schools`

In [1]:
import pandas as pd
import numpy as np

In [5]:
bos = pd.read_csv('../datasets/buildbps.csv')
bos

SMMA_Identifier SMMA_Only_For_Map                       BPS_School_Name  \
0               031               NaN              Adams, Samuel Elementary   
1               078               NaN    Alighieri, Dante Montessori School   
2              045A               NaN            Another Course to College*   
3               012               NaN  Baldwin Early Learning Pilot Academy   
4               087               NaN             Bates, Phineas Elementary   
..              ...               ...                                   ...   
136            011B               NaN                 West Roxbury Academy*   
137            047B               NaN      West Zone Early Learning Center*   
138             100               NaN          Winship, F. Lyman Elementary   
139             073               NaN             Winthrop, John Elementary   
140             044               NaN    Young Achievers Science & Math K-8   

            BPS_Historical_Name SMMA_Abbreviated_Name  \
0                         Adams                 Adams   
1                     Alighieri             Alighieri   
2                          Taft                  ACC*   
3                   Baldwin ELC               Baldwin   
4                         Bates                 Bates   
..                          ...                   ...   
136    West Roxbury High School         West Roxbury*   
137  Hennigan Elementary School            West Zone*   
138                     Winship               Winship   
139         Winthrop Elementary              Winthrop   
140                   Lewenberg       Young Achievers   

                               BPS_Address BRA_Neighborhood  SMMA_latitude  \
0     165 Webster St East Boston, MA 02128      East Boston      42.365553   
1     37 Gove Street East Boston, MA 02128      East Boston      42.371565   
2      20 Warren Street Brighton, MA 02135          Allston      42.350354   
3          121 Corey Rd Brighton, MA 02135         Brighton      42.342037   
4        426 Beech St Roslindale, MA 02131       Roslindale      42.277663   
..                                     ...              ...            ...   
136  1205 VFW Pkwy, West Roxbury, MA 02132     West Roxbury      42.282206   
137   200 Heath St Jamaica Plain, MA 02130    Jamaica Plain      42.326026   
138       54 Dighton St Brighton, MA 02135         Brighton      42.347723   
139   35 Brookford St Dorchester, MA 02125          Roxbury      42.318387   
140        20 Outlook Rd Mattapan, MA02126         Mattapan      42.281976   

     SMMA_longitude      SMMA_Typology  ... SMMA_EA_K8_Adequacy_Cafeteria  \
0        -71.034917  Elementary School  ...                           NaN   
1        -71.037608  Elementary School  ...                           NaN   
2        -71.145582        High School  ...                           NaN   
3        -71.140529     Early Learning  ...                           NaN   
4        -71.135353  Elementary School  ...                           NaN   
..              ...                ...  ...                           ...   
136      -71.174549        High School  ...                           NaN   
137      -71.106821     Early Learning  ...                           NaN   
138      -71.154965  Elementary School  ...                           NaN   
139      -71.075341  Elementary School  ...                           NaN   
140      -71.093909                K-8  ...                          Poor   

    SMMA_EA_K8_Adequacy_Stage SMMA_EA_K8_Adequacy_Medical  \
0                         NaN                         NaN   
1                         NaN                         NaN   
2                         NaN                         NaN   
3                         NaN                         NaN   
4                         NaN                         NaN   
..                        ...                         ...   
136                       NaN                         NaN   
137                 

### Columns

One of the greatest challenges in this notebook is going to be examining the columns, filling `NaaNs`, and choosing which columns should be dropped from this dataset. After all, there are **251** columns worth of information for just 141 schools.

One cannot view all the columns in this dataset simply by running the command `bos.columns` -- there are simply too many to display. Let's look at all the columns in batches of 50.

In [9]:
bos.columns[:50]

Index(['SMMA_Identifier', 'SMMA_Only_For_Map', 'BPS_School_Name',
       'BPS_Historical_Name', 'SMMA_Abbreviated_Name', 'BPS_Address',
       'BRA_Neighborhood', 'SMMA_latitude', 'SMMA_longitude', 'SMMA_Typology',
       'SMMA_Educational_Program', 'BPS_Grades_Offered', 'BPS_Property_Status',
       'BPS_Total_GSF', 'BPS_GSF', 'SMMA_Site_SF', 'SMMA_Site_Acres',
       'BPS_Year_Founded ', 'BPS_Year_Built', 'MSBA_Year_Reno', 'SMMA_Era',
       'BPS_Plans', 'BPS_Principal', 'BPS_Open', 'BPS_Close', 'DOE_ID',
       'DOE_Total', 'DOE_PK', 'DOE_K', 'DOE_1', 'DOE_2', 'DOE_3', 'DOE_4',
       'DOE_5', 'DOE_6', 'DOE_7', 'DOE_8', 'DOE_9', 'DOE_10', 'DOE_11',
       'DOE_12', 'DOE_SP', 'GSF / Student Value', 'BPS_KWH',
       ' BPS_Electric_Bill ', 'BPS_Therms', ' BPS_Gas_Bill ',
       'BPS_Water_ft_cubic', ' BPS_Water_Bill ', ' BPS_Total_Energy_Cost '],
      dtype='object')

In [13]:
bos.columns[50:100]

Index(['Climate_Priority', 'Climate_Risk', 'SMMA_FA_Visit_Date',
       'SMMA_FA_Status', 'SMMA_FA_School_Enrollment',
       'SMMA_FA_Enrollment_Building', 'SMMA_FA_Major_Investments',
       'SMMA_FA_Existing_PV', 'SMMA_FA_space_Solar', 'SMMA_FA_Roof_Membrane',
       'SMMA_FA_Facade', 'SMMA_FA_Windows', 'SMMA_FA_Boilers',
       'SMMA_FA_Heating_Distribution_Systems',
       'SMMA_FA_Ventilation_Distribution_Systems',
       'SMMA_FA_Electrical_Service', 'SMMA_FA_Means_Egress',
       'SMMA_FA_FP_Sprinklers', 'SMMA_FA_Life_Safety',
       'SMMA_FA_Security_Entry', 'SMMA_FA_Lighting_Quantity_Control',
       'SMMA_FA_Toilets_Fixtures', 'SMMA_FA_Plumbing_Distribution_Systems',
       'SMMA_FA_Accessibility', 'SMMA_FA_Deterioration_Roof',
       'SMMA_FA_Deterioration_Floor', 'SMMA_FA_Deterioration_Walls_Columns',
       'SMMA_FA_Deterioration_Foundations', 'SMMA_FA_Deterioration_Facade',
       'SMMA_FA_lateral_system_identifiable',
       'SMMA_FA_Overall_Building_Condition', 'SMMA_F

In [15]:
bos.columns[100:150]

Index(['SMMA_FA_Walls_Slopes', 'SMMA_FA_Overall_Site_Condition',
       'SMMA_FA_MHC_Archaeological_Assets', 'SMMA_FA_Mass_Transit',
       'SMMA_FA_Bikable', 'SMMA_FA_Walkable', 'SMMA_FA_Overall_Community_Site',
       'SMMA_FA_Site_suitability_school', 'Visit_Date_Date', 'SMMA_ES_Status',
       'SMMA_EA_Building_originally_designed',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_1',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_3',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_5',
       'SMMA_EA_grade_configuration_best_suited_PreK_to_6',
       'SMMA_EA_grade_configuration_best_suited_4_to_6',
       'SMMA_EA_grade_configuration_best_suited_6_to_8',
       'SMMA_EA_grade_configuration_best_suited_7_to_8',
       'SMMA_EA_grade_configuration_best_suited_6_to_12',
       'SMMA_EA_grade_configuration_best_suited_7_to_12',
       'SMMA_EA_grade_configuration_best_suited_9_to_12',
       'SMMA_EA_Building_Ventilation', 'SMMA_EA_Building_Natural_Daylighting

In [16]:
bos.columns[150:200]

Index(['SMMA_EA_ES_Adequacy_Music', 'SMMA_EA_ES_Adequacy_Gymnasium',
       'SMMA_EA_ES_Adequacy_Media_Center', 'SMMA_EA_ES_Adequacy_Cafeteria',
       'SMMA_EA_ES_Adequacy_Stage', 'SMMA_EA_ES_Adequacy_Medical',
       'SMMA_EA_ES_Adequacy_Administration', 'SMMA_EA_ES_Adequacy_Custodial',
       'SMMA_EA_ES_Adequacy_Network', 'SMMA_EA_ES_Adequacy_Other_1',
       'SMMA_EA_ES_Adequacy_Other_2', 'SMMA_EA_ES_Adequacy_Other_3',
       'SMMA_EA_ES_Overall_EFE_spaces', 'SMMA_EA_MS_Adequacy_Classroom',
       'SMMA_EA_MS_Adequacy_Science', 'SMMA_EA_MS_Adequacy_SPED_SS',
       'SMMA_EA_MS_Adequacy_SPED_Resource', 'SMMA_EA_MS_Adequacy_Art',
       'SMMA_EA_MS_Adequacy_Music', 'SMMA_EA_MS_Adequacy_Vocations',
       'SMMA_EA_MS_Adequacy_Gymnasium', 'SMMA_EA_MS_Adequacy_Media_Center',
       'SMMA_EA_MS_Adequacy_Cafeteria', 'SMMA_EA_MS_Adequacy_Stage',
       'SMMA_EA_MS_Adequacy_Medical', 'SMMA_EA_MS_Adequacy_Administration',
       'SMMA_EA_MS_Adequacy_Custodial', 'SMMA_EA_MS_Adequacy_Network'

In [17]:
bos.columns[200:252]

Index(['SMMA_EA_HS712_Adequacy_Network', 'SMMA_EA_HS712_Adequacy_Other_1',
       'SMMA_EA_HS712_Adequacy_Other_2', 'SMMA_EA_HS712_Adequacy_Other_3',
       'SMMA_EA_HS712_Overall_EFE_spaces', 'SMMA_EA_HS912_Adequacy_Classroom',
       'SMMA_EA_HS912_Adequacy_TeacherPlanning',
       'SMMA_EA_HS912_Adequacy_Small_Group', 'SMMA_EA_HS912_Adequacy_Science',
       'SMMA_EA_HS912_Adequacy_SPED_SS',
       'SMMA_EA_HS912_Adequacy_SPED_Resource', 'SMMA_EA_HS912_Adequacy_Art',
       'SMMA_EA_HS912_Adequacy_Music', 'SMMA_EA_HS912_Adequacy_Vocations',
       'SMMA_EA_HS912_Adequacy_Gymnasium',
       'SMMA_EA_HS912_Adequacy_GynasiumAlternatives',
       'SMMA_EA_HS912_Adequacy_Media_Center',
       'SMMA_EA_HS912_Adequacy_Cafeteria', 'SMMA_EA_HS912_Adequacy_Auditorium',
       'SMMA_EA_HS912_Adequacy_Stage', 'SMMA_EA_HS912_Adequacy_Medical',
       'SMMA_EA_HS912_Adequacy_Administration',
       'SMMA_EA_HS912_Adequacy_Custodial', 'SMMA_EA_HS912_Adequacy_Network',
       'SMMA_EA_HS912_Adequac

#### Where do we begin?

One good place to be would be with the `NaaNs`. If there are any columns with a large about of data missing (i.e., 50% of more) -- dropping these columns from the dataset could be a good place to start.

In [39]:
# Let's look at 100 columns with the most NaaNs
bos.isna().sum().sort_values(ascending = False)[:100]

SMMA_EA_HS712_Adequacy_Network                    141
SMMA_EA_grade_configuration_best_suited_4_to_6    141
SMMA_EA_MS_Adequacy_Network                       141
SMMA_EA_HS712_Adequacy_Other_3                    141
SMMA_EA_K8_Adequacy_Network                       140
                                                 ... 
SMMA_EA_K8_Adequacy_Media_Center                  105
SMMA_EA_K8_Adequacy_Cafeteria                     105
SMMA_EA_K8_Adequacy_Stage                         105
SMMA_EA_K8_Adequacy_SPED_SS                       105
SMMA_EA_K8_Adequacy_Medical                       105
Length: 100, dtype: int64

100 columns in the dataset have **75% or more** missing data! The data in these columns may have been valuable, but I wouldn't feel comfortable filling `NaaNs` with any values (i.e., mean value of the column) with so much data missing. Let's drop these columns from the dataset.

In [40]:
[column for column in bos.columns if bos[column].isna().sum() >= 105]

['SMMA_Only_For_Map',
 'MSBA_Year_Reno',
 'SMMA_EA_grade_configuration_best_suited_4_to_6',
 'SMMA_EA_grade_configuration_best_suited_6_to_8',
 'SMMA_EA_grade_configuration_best_suited_7_to_8',
 'SMMA_EA_grade_configuration_best_suited_6_to_12',
 'SMMA_EA_grade_configuration_best_suited_7_to_12',
 'SMMA_EA_grade_configuration_best_suited_9_to_12',
 'SMMA_EA_ES_Adequacy_Custodial',
 'SMMA_EA_ES_Adequacy_Network',
 'SMMA_EA_ES_Adequacy_Other_1',
 'SMMA_EA_ES_Adequacy_Other_2',
 'SMMA_EA_ES_Adequacy_Other_3',
 'SMMA_EA_MS_Adequacy_Classroom',
 'SMMA_EA_MS_Adequacy_Science',
 'SMMA_EA_MS_Adequacy_SPED_SS',
 'SMMA_EA_MS_Adequacy_SPED_Resource',
 'SMMA_EA_MS_Adequacy_Art',
 'SMMA_EA_MS_Adequacy_Music',
 'SMMA_EA_MS_Adequacy_Vocations',
 'SMMA_EA_MS_Adequacy_Gymnasium',
 'SMMA_EA_MS_Adequacy_Media_Center',
 'SMMA_EA_MS_Adequacy_Cafeteria',
 'SMMA_EA_MS_Adequacy_Stage',
 'SMMA_EA_MS_Adequacy_Medical',
 'SMMA_EA_MS_Adequacy_Administration',
 'SMMA_EA_MS_Adequacy_Custodial',
 'SMMA_EA_MS_Adequac

In [43]:
# Drop columns with over 105 NaaNs from dataset
bos = bos.drop(columns = [column for column in bos.columns if bos[column].isna().sum() >= 105])

# Check work
bos.isna().sum().sort_values(ascending = False)

SMMA_EA_K8_Overall_EFE_spaces                        104
SMMA_EA_grade_configuration_best_suited_PreK_to_6    104
Climate_Risk                                          99
Climate_Priority                                      97
SMMA_EA_ES_Adequacy_Art                               92
                                                    ... 
SMMA_FA_Enrollment_Building                            0
SMMA_FA_School_Enrollment                              0
SMMA_FA_Status                                         0
 BPS_Total_Energy_Cost                                 0
SMMA_Identifier                                        0
Length: 151, dtype: int64

Even after dropping 100 columns from our dataset that were riddled with `NaaNs`, we still have 151 columns, some of which still have over 50% `NaaNs` in them (see above). Let's take a look at ant columns with over `50%` of their values missing from the dataset.

In [44]:
[column for column in bos.columns if bos[column].isna().sum() >= bos.shape[0] / 2]

['Climate_Priority',
 'Climate_Risk',
 'SMMA_EA_grade_configuration_best_suited_PreK_to_5',
 'SMMA_EA_grade_configuration_best_suited_PreK_to_6',
 'SMMA_EA_ES_Adequacy_PreK',
 'SMMA_EA_ES_Adequacy_Kindergarten',
 'SMMA_EA_ES_Adequacy_Classroom',
 'SMMA_EA_ES_Adequacy_SPED_SS',
 'SMMA_EA_ES_Adequacy_SPED_Resource',
 'SMMA_EA_ES_Adequacy_Art',
 'SMMA_EA_ES_Adequacy_Music',
 'SMMA_EA_ES_Adequacy_Gymnasium',
 'SMMA_EA_ES_Adequacy_Media_Center',
 'SMMA_EA_ES_Adequacy_Cafeteria',
 'SMMA_EA_ES_Adequacy_Stage',
 'SMMA_EA_ES_Adequacy_Medical',
 'SMMA_EA_ES_Adequacy_Administration',
 'SMMA_EA_ES_Overall_EFE_spaces',
 'SMMA_EA_K8_Overall_EFE_spaces']

In [46]:
# Drop columns with over 50% NaaNs from dataset
bos = bos.drop(columns = [column for column in bos.columns if bos[column].isna().sum() >= bos.shape[0] / 2])

# Check work
bos.isna().sum().sort_values(ascending = False)

SMMA_EA_grade_configuration_best_suited_PreK_to_3    64
SMMA_EA_grade_configuration_best_suited_PreK_to_1    56
SMMA_EA_Building_originally_designed                 18
DOE_SP                                               17
BPS_Year_Founded                                     17
                                                     ..
SMMA_FA_Overall_Building_Condition                    0
SMMA_FA_Overall_Site_Condition                        0
SMMA_FA_Overall_Community_Site                        0
SMMA_EA_Overall_EFE_learning_environments             0
SMMA_Identifier                                       0
Length: 132, dtype: int64

Ok...last command in which `NaaNs`, I promise.

After looking at the columns that remain in our dataset, we can see that there are still two columns with `64` and `56` `NaaN` values. After that, the next highest `NaaN` count is `18`, which is much more managable.

Let's drop those other 2 columns from our dataset.

In [47]:
bos.drop(columns = ['SMMA_EA_grade_configuration_best_suited_PreK_to_3',
                    'SMMA_EA_grade_configuration_best_suited_PreK_to_1'],
         inplace = True)